### ISSUE TITLE
8.4-Evaluation of GAN

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#54]
Train GAN 最難的其實是要拿 GAN 來生成文字

如果要生成一段文字, 那可能會有一個 Sequence To Sequence 的 Model
之前是用 Decoder 產生一段文字, 而現在Sequence To Sequence 的 Model 就是我們的 Generator 
在 Transformer 的時候是 Decoder , 在 GAN 裡面它就扮演了 Generator 的角色

跟原來在影像上的 GAN 有什麼不同呢?

就演算法來看可能沒有太大的不同
訓練一個 Discriminator, Discriminator 把這段文字讀進去, 判斷說這段文字是真正的文字還是機器產生出來的文字
而 Decoder(Generator) 就是調整 Generator 的參數想辦法去騙過 Discriminator

```
然而無法用 Gradient Descent 去 Train 你的 Decoder
當 Decoder 的參數有一點變化的時候Discriminator 輸出是沒有改變的, 所以你根本就沒有辦法算微分
```

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/b6da3cfd-2ab9-474f-ba24-6ab421c7a2c2)

所謂的 Gradient Descent其實就是某一個參數有變化的時候, 對你的目標造成了多大的影響

假設我們改變了 Decoder 的參數 (也就是 Generator 的參數)

Decoder 的參數有一點小小的變化->輸出的 Distribution也會有小小的變化
因為這個變化很小-> 分數最大的那個 Token 是沒有改變的->不會影響output的Token

________
### Token
Token = 你現在處理產生 Sequence 的單位(使用者自訂)
在產生一個中文的句子的時候我們是每次產生一個 Character(一個方塊字), 方塊字就是我們的 Token

在處理英文的時候
你每次產生一個英文的字母,那字母就是你的 Token 
假設你一次是產生一個英文的詞, 英文的詞和詞之間是以空白分開的, 那詞就是你的 Token
___________

Distribution 只有小小的變化->分數最大的那個 Token 是同一個-> Discriminator 輸出的分數一樣

之前提過如果遇到不能用 Gradient Descent Train 的問題, 可以使用Reinforcement Learning 硬做
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/b3b3f1a5-95c2-47e8-8877-0539d83f3bcf)

所以你確實可以用 Reinforcement Learning來 Train 你的 Generator 
然而Reinforcement Learning 是以難 Train 而聞名 , GAN 也是以難 Train 而聞名
這樣的東西加在一起就會非常非常地難訓練
所以要用 GAN 產生一段文字, 過去一直被認為通常你需要先做 Pretrain(無法直接train, 要先做其他前置工作)

直到 Train Language GANs Form Scrach這篇 paper問世(Form Scrach =不用 Pretrain 的意思)
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8a92bcb4-3f11-4667-92bf-694da06f559a)

它可以直接從隨機的初始化參數開始train
它怎麼做到的呢?最關鍵的就是爆收爆調 Hyperparameter跟一大堆的 Tips 融合(不細講)

其實有關 Generative 的 Model 不是只有 GAN 而已, 還有 VAE 或 FLOW-Based Model 
但GAN 的 Performance 是比較好的 

_____________
# Evaluation of GAN

怎麼評估一個 Generator 的好壞並沒有那麼容易
那最直覺的做法也許是找人來看(人工的方式)這個 Generator 產生出來的圖片到底像不像動畫的人物
人們剛開始研究Generative 這樣的技術的時候
很長一段時間沒有好的 Major 評估 Generator 的好壞
都是人眼看, 然後直接用吹的這樣
(在 Paper 最後就放幾張圖說 你看這個我覺得應該是比文獻上目前結果都還要好)

但是沒有衡量的指標顯然是不行的
完全用人來看顯然有很多的問題, 比如說不客觀, 不穩定等等諸多的問題 

如果針對特定的一些任務是有辦法設計一些方法衡量

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/ee83794c-59e0-4d6c-93e6-adab77885e83)

舉例來說:在作業 6 裡面就是要叫大家生成二次元人物的頭像
作業裡面一個評估的標準, 就是我們跑一個動畫人物人臉偵測的系統
如果你提供 1000 張圖片裡面 抓到 900 個人臉
跟提供 1000 張圖片抓到 300 個人臉
那顯然 900 個人臉的那一個 Generator 它做出來的結果是比較好的

那如果是跟一般的 Case 我們怎麼知道它做得好不好呢?

一般的 Case 我隨便訓練了一個 Generator, 它不一定是產生動畫人物的
因為它可以產生別的, 像是它專門產生貓, 專門產生狗, 專門產生斑馬等等

一樣跑一個影像的分類系統
把GAN 產生出來的圖片丟到一個的影像的分類系統裡面
影像分類系統輸入是一張圖片輸出y 
P ( c│y ) 是一個機率的分布
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/41991678-19a3-438b-a9ec-182dbea86c4f)

這個機率的分布如果越集中, 就代表說現在產生的圖片可能越好

雖然我們不知道這邊產生的圖片是貓還是狗還是斑馬(我們不知道它是什麼)
但是如果影像分類系統輸出來的分布非常集中, 代表影像分類系統它非常肯定它現在看到什麼樣的東西

如果你產生出來的圖片是一個四不像(根本看不出是什麼動物)
那影像辨識系統就會非常地困惑, 它產生出來的這個機率分布就會非常地平坦
如果平均分布的話, 那就代表說你的 GAN產生出來的圖片可能是比較奇怪的

但是光用這個做法你會被一個叫做 Mode Collapse 的問題騙過去
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/f1e0a0ed-407f-4c15-a893-12b9e393b379)

### Mode Collapse
Generative Model 輸出來的圖片來來去去就是那幾張 = 輸出data高度集中

當 Generator 學會產生可以騙過 Discriminator的圖片以後
Discriminator 沒辦法看出說這樣子的圖片是假的( Discriminator 的盲點)
Generator 抓到這個盲點就硬打一發, 狂生成類似這種的圖片就發生 Mode Collapse 的狀況

Q:怎麼避免 Mode Collapse 的狀況呢?
A:李認為今天其實還沒有一個非常好的解答

### Mode Dropping 

```
Mode Dropping=你的產生出來的資料, 只有真實資料的一部分
```

單純看產生出來的資料, 你可能會覺得還不錯, 它的這個多樣性夠
但你不知道說真實的資料, 它的多樣性的分布其實是更大的

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/c89cf63a-062e-48c4-8563-0aea85f9f827)

真實的例子 -Train 了人臉生成 GAN

那它在某一個 Iteration 的時候
它的 Generator 產生出這些人臉你會覺得說沒有問題(人臉的多樣性足夠)
但如果你再看下一個 Iteration會發現膚色偏黃
你沒弄好人家都覺得你的 Generator 有種族歧視

事實上今天這些非常好的 GAN (BGAN Progress GAN BGAN Progress GAN )
可以產生非常真實人臉這些 GAN 可能還是依舊有 Mode Dropping 的問題
_____________

### Generator Diversity

Generator 它產生出來的圖片到底多樣性夠不夠?

過去做法

把你的 Generator 產生 1000 張圖片都丟到 Image Classify 裡面
每張圖片都會給我們一個 Distribution

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/055d6244-800a-4c8c-9514-a7031771f6d3)

把所有的 Distribution 平均起來
如果平均的 Distribution 非常集中就代表現在多樣性不夠

如果任何圖片丟進去你的影像分類系統都說 是看到 Class 2
那代表說 每一張圖片也許都蠻像的=你的多樣性是不夠的

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/9fdcfde0-06c0-4898-abca-95f06a5b51a6)

你的 Generator 產生出來的圖片丟到 Image Classifier 的時候
它產生出來的輸出的分布都非常地不同->你平均完結果是非常平坦的->也許多樣性足夠

當我們用這個 Image 的 Classifier 來做評估的時候
Quality 是只看一張圖片, 丟到 Classifier 的時候分布有沒有非常地集中
Diversity 看的是一堆圖片, 它一堆圖片分布的平均 

### Inception Score

那過去有一個非常常被使用的分數- Inception Score = IS 

用 Inception Network 量一下 Quality以及Diversity 
如果 Quality 高, Diversity 又大-> Inception Score 大

不過在作業裡面我們並不會用 Inception Score 
假設把你產生出來的二次元人物丟到 Inception Net 裡面
你可能 Diversity 很大 (產生不同的髮色 產生不同的眼睛的顏色的人物)
但是對 Inception Network 來說 它都是人臉(所以你可能算出來 Diversity 其實是小的)

所以在我們的作業中會採取另外一個 Evaluation 的 Measure :Fréchet 的 Inception Distance  = FID

把產生出來的二次元的人物丟到 Inception Net 裡面
不取最後的類別只取這個 Inception Network 最後一層的這個 Hidden Layer 的輸出
也就是進入 Softmax 之前的一個高維度向量(圖片)

所有紅色的點代表你把真正的圖片丟到 Inception Network 以後拿出來的向量
藍色的點是你自己的 GAN 你自己的 Generator 產生出來的圖片

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/0a6414d6-a1f2-42b2-af2d-2b6f2aef611b)

假設真實的圖片跟產生出來的圖片它們都是 Gaussians Distribution
計算這兩個 Gaussians Distribution 之間的Fréchet 的 Distance 
(在作業裡面 Judge Systerm 會幫大家算好)

Fréchet 的 Distance 距離就是越小越好 (代表這兩組圖片越接近)

第一個問題是 不一定是 Gaussians Distribution
第二個問題是 如果你要準確的得到你的 Network 它的分布那你可能需要產生大量的 Sample 才能做到
如果只看FID 很怪, 因為不一定是Gaussians Distribution我們卻這樣假設
所以我們是同時看 FID 跟偵測出來的人臉的數目這兩個指標

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/8011deb3-6ee4-450d-9d09-faf03ce6b2eb)

有一篇 Paper 叫做Are GANs Created Equal A Large Scale Study
各式各樣的 GAN 每一種 GAN它都用不同的 Random Seed 去跑很多次
不同的 Random Seed 去跑會都得到一個分布, 四個不同的資料庫上面得到的結果

其實剛才那些 Measure 也並沒有完全解決 GAN Evaluation 的問題

你不知道怎麼回事 訓練了一個 Generator 它產生出來的 Data 跟你的真實資料一模一樣
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/72a405e4-20de-4d6e-a90f-52b782d91473)

如果你不知道真實資料長什麼樣子
你光看這個 Generator 的輸出你會覺得太棒了

 FID 算出來 一定是非常小的, 但問題是這個並不是你要的

如果它產生出來的圖片都跟資料庫裡面的訓練資料的一模一樣
訓練資料就在你手上, 直接從訓練資料裡面 Sample 一些 Image 出來不是更好
我們 Train Generator 其實是希望它產生新的圖片啊(訓練資料裡面沒有的人臉啊)

光要如何評估 一個 Generator 做得好不好這件事情 都是一個可以研究的題目
這邊放了一篇相關的文章裡面就列舉了二十幾種GAN Generator 的評估的方式

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/fec0228c-4f0c-42e0-a873-2c0a229f656d)
